Refactor BulletCursorOverlay using Clone DOM Approach

cindmichelle opened the pull request:
Fix #3162 

## Results
### with saving cursor thought ref in ministore approach 

https://github.com/user-attachments/assets/3e43e17c-3098-4851-9208-6cf87b632ca6

---
### indicate cursor thought with `data-cursor=true` DOM attribute approach (latest)

https://github.com/user-attachments/assets/db1466e6-65ac-4dd4-aa9c-247285b73707

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,290 +1,86 @@
+      // Find and hide the thought-annotation element
+      const thoughtAnnotation = nestedChild.querySelector('div[aria-label="thought-annotation"]') as HTMLElement
+      if (thoughtAnnotation) {
+        thoughtAnnotation.style.visibility = 'hidden'
```
I'd like to better understand why you are setting cloned descendant elements to visibility `hidden` in `BulletCursorOverlay` rather than in `TreeNode` when the node is cloned. It seems like something to do during cloning. Could you explain?

cindmichelle replied:
I didnt set `visibility: hidden` on the TreeNode level before because it would affect everything (including the cursor overlay SVG). But I just got the idea to set `visibility: visible` back on the CursorOverlay component itself, and it seems to work. Thanks for sparking the idea and helping me explore the implementation further!

raineorshine replied:
Also, can the nodes be removed completely instead of hiding them?

cindmichelle replied:
At first I was only going to clone the necessary elements, but then I thought if someone adds or removes something under the TreeNode, it could affect the bullet position again. So I decided to just clone the whole thing to avoid having to adjust it manually later on. Whats your take on this trade-off?

raineorshine replied:
> I think my explanation above was a bit confusing, so let me spell out my concern about **only cloning the necessary parts**. This also depends on how we do it, either we _pick up only what’s necessary_, or we _clone everything and then remove what’s unnecessary._

We clone everything and then remove what's unnecessary. My question is more about _when_ to remove what's unnecessary. Do we do it immediately after the clone, or wait until the clone is rendered? I'm advocating for the former if possible.

> On the flip side, if we clone everything and try to strip out the unnecessary parts, then when someone adds a new element in the original component that actually doesn't affect the bullet position, we’d still have to manually remove it from the clone again.

I don't totally understand this. Can you explain what you mean by adding a new element? Are you talking about the user or a developer, and what kind of new element?

> I’m also curious, if we clone and includes the unnecessary elements when render, is that going to hurt performance?

I don't think it'll have a big effect on performance. It's more about trying to encapsulate this functionality better.

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,290 +1,86 @@
+        thoughtAnnotation.style.visibility = 'hidden'
+      }
+
+      const staticThought = nestedChild.querySelector('div[aria-label="static-thought"]') as HTMLElement
```
This is one of many tight couplings to the DOM. I'd like to consider how we might be able to reduce the surface area to make this more encapsulated and less fragile.

cindmichelle replied:
yep we can avoiding this pattern by implementing this comment now https://github.com/cybersemics/em/pull/3211/files#r2319893043

raineorshine commented on src/stores/cursor.ts:
```diff
@@ -0,0 +1,58 @@
+}) => {
+  const oldState = ministore.getState()
+  ministore.update({
+    ...oldState,
```
`update` should already handle partial updates, so you don't need to merge in `...oldState`.

raineorshine commented on src/components/TreeNode.tsx:
```diff
@@ -210,9 +214,28 @@ const TreeNode = ({
+        }
+      }
+
+      cursorStore.onChange({
+        treeNode: clonedOuterDiv,
+      })
```
What's the advantage of going through the ministore back channel rather than appending it directly?

The ministore adds complexity, and we already have a lot of direct DOM couplings, but maybe there's something else I'm not considering.

cindmichelle replied:
We cant append it directly (from TreeNode.tsx) because the thing we want to append is still a React element, whereas `parent.appendChild(child)` only works with DOM nodes/refs. so we use `ReactDOM.createPortal()` instead

Since portals are declarative, we place `createPortal()` inside the BulletCursorOverlay component (as it is now). The portal also needs a target to mount into, so I’m storing `targetSvg` in state inside `BulletCursorOverlay`.

For passing the cloned element around, alternatives to the ministore would be React Context or prop drilling, but functionally they’re equivalent here (just different transport mechanisms).

raineorshine replied:
Yes, I'm interested in that. It's direct DOM manipulation, but we already have so much of that anyway.

If it was a clean portal with just the mounting node, that would be one thing. But there are so many DOM nodes that are being referenced for cloning and removing, I think the declarative benefits of the portal are somewhat lost anyway.

cindmichelle commented:
Hi @raineorshine, just updated this PR with direct manipulation approach (querying the DOM with `data-cursor=true` to get the latest cursor node). 

Right now I’m using 2× requestAnimationFrame to delay the querying DOM elements before the `data-cursor` attr switches to the new cursor thought, and also to capture the final position when in table view. 

For Android and iOS, I decided to add one extra rAF since I noticed a slight position shift when the virtual keyboard appears, which makes the position without the keyboard invalid. I also saw a similar issue when switching from normal view to table view, where the position was incorrect. So I added one more rAF to ensure DOM cloning happens after the browser knows whether the virtual keyboard has appeared, and to fix the table view transition issue.


https://github.com/user-attachments/assets/db1466e6-65ac-4dd4-aa9c-247285b73707

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,375 +1,157 @@
-    width: isTableCol1 ? width : `calc(100% - ${x}px + 1em + 10px)`,
-    textAlign: isTableCol1 ? ('right' as const) : undefined,
-  }
+    return { simplePath: path, thoughtId: id, childrenThoughts: children, isInContextView: contextView }
```
Since you are returning an object literal, this will return a new object reference every time, regardless of whether any of the values have changed, i.e. `({}) !== ({})`. This has the unfortunate effect of causing the component to re-render on every `State` change.

To do a deep equal comparison and only re-render when a value changes, pass lodash's `isEqual` as the equality function:

```
useSelector(fn, isEqual)
```

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,375 +1,157 @@
+  const updateCursorNode = useCallback(() => {
+    if (!containerRef.current || !cursor) return
+
+    // Find the cursor TreeNode using the data-cursor attribute
```
A bit more explanation needed here to explain why two `data-cursor` nodes would exist in the DOM at the same time.

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,375 +1,157 @@
+      if (svg) {
+        // Replace the SVG contents
+
+        const svgInnerHTML = `<g style="visibility: visible; will-change: transform;"><ellipse ry="${bulletOverlayRadius}" rx="${bulletOverlayRadius}" cy="300" cx="300" class="stk_highlight fill-opacity_0.25 fill_fg"></ellipse></g>`
```
Let's assign this directly rather than creating a separate variable for it.

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,375 +1,157 @@
+
+    const container = containerRef.current
+
+    // Copy all attributes from the cursor TreeNode to the container (except data-cursor)
```
Can you add a comment explaining which attributes are copied and why this is necessary?

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,375 +1,157 @@
+
+    // Clear existing content and copy the nested child from cursor TreeNode
+    container.innerHTML = ''
+    container.style.visibility = 'hidden'
```
What is the reason for setting visibility to `hidden` on the container?

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,375 +1,157 @@
+        const svgInnerHTML = `<g style="visibility: visible; will-change: transform;"><ellipse ry="${bulletOverlayRadius}" rx="${bulletOverlayRadius}" cy="300" cx="300" class="stk_highlight fill-opacity_0.25 fill_fg"></ellipse></g>`
+        svg.innerHTML = svgInnerHTML
+
+        svg.style.visibility = 'visible'
```
What is the reason for setting visibility to `visible` on the svg?

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -1,375 +1,157 @@
+        })
+      : false
+
+  const prevDomElementRef = React.useRef<HTMLElement | null>(null)
```
Let's add a JSDOC explaining the purpose of this ref.
